pkg/cmd/pulumi/config/config.go

```diff
@@ -808,6 +844,60 @@ func newConfigSetAllCmd(stack *string) *cobra.Command {
 				}
 			}
 
+			if jsonArg != "" {
+				jsonConfig := make(map[string]configValueJSON)
+				err := json.Unmarshal([]byte(jsonArg), &jsonConfig)
+				if err != nil {
+					return fmt.Errorf("could not parse --json argument: %w", err)
+				}
+
+				for jsonKey, jsonValue := range jsonConfig {
+					// Every key in the JSON must have a value. While `pulumi config --json` may produce empty values for secrets
+					// where --show-secrets is not passed, we don't allow setting those same (missing) values back into
+					// configuration.
+					if jsonValue.Value == nil {
+						return fmt.Errorf("value for --json object key %q is nil", jsonKey)
+					}
+
+					key, err := ParseConfigKey(ws, jsonKey, false /*path*/)
+					if err != nil {
+						return fmt.Errorf("invalid --json object key %q: %w", jsonKey, err)
+					}
+
+					var value config.Value
+					var path bool
+					if jsonValue.ObjectValue != nil {
+						// If we have an ObjectValue, Value will be the stringified version of the object. We'll thus operate on
+						// ObjectValue directly.
+						path = true
+						if jsonValue.Secret {
+							enc, err := encrypt(*jsonValue.Value)
+							if err != nil {
+								return err
+							}
+
+							value = config.NewSecureObjectValue(enc)
+						} else {
+							value = config.NewObjectValue(*jsonValue.Value)
+						}
+					} else if jsonValue.Secret {
+						enc, err := encrypt(*jsonValue.Value)
+						if err != nil {
+							return err
+						}
+
+						value = config.NewSecureValue(enc)
+					} else {
+						value = config.NewValue(*jsonValue.Value)
+					}
+
+					err = ps.Config.Set(key, value, path)
+					if err != nil {
+						return fmt.Errorf("could not set --json config for %q: %w", jsonKey, err)
+					}
+				}
+			}
```

Comment:
This is starting to hang a little far to the right.  Can we factor it out into its own function?

Reply:
Good shout -- will do

Reply:
I'd made an attempt to factor out the JSON block, didn't like the loss of clarity.  You'd already approved the PR, so I took your comment as a nit.  My mistake.  On the merits though, I wouldn't change it.

Reply:
Any comment deserves at least an explanation of why it hasn't been addressed instead of being silently ignored.  If I approve the PR it's okay to go ahead and merge it if you disagree with my comment, but I'd like to know why, and that it wasn't just ignored.

Disagreeing and merging as here is fine though, thanks for the explanation.